src/librustc_mir/hair/pattern/mod.rs

```diff
@@ -179,7 +179,12 @@ impl<'tcx> PatternTypeProjection<'tcx> {
         annotations: &mut CanonicalUserTypeAnnotations<'tcx>,
         span: Span,
     ) -> UserTypeProjection<'tcx> {
-        let annotation_index = annotations.push((span, self.base));
+        let annotation = (span, self.base);
+        let annotation_index = if let Some(index) = annotations.position(|a| a == &annotation) {
```

Comment:
I'm not convinced this is correct. Does it include some kind of link to the user's source? For example, what if we have:

```rust
let x: Vec<_> = ...;
let y: Vec<_> = ...;
```

would those two instances of `Vec<_>` wind up as the same canonical user type annotation?

Reply:
In further discussion on [zulip](https://rust-lang.zulipchat.com/#narrow/stream/122657-wg-nll/subject/.2354943.20user.20type.20annot.20in.20unreachable.20code/near/152252966), @davidtwco provided a different path towards a more limited avoidance of duplication that will also resolve my issues on #55748